src/bun.js/test/expect.zig

```diff
@@ -3707,10 +3803,197 @@ pub const Expect = struct {
+            if (message_result.isAnyError()) {
+                globalObject.throwValue(message_result);
+                return false;
+            }
+
+            const message = bun.String.init(message_result.toString(globalObject).getZigString(globalObject));
+            const fmt = comptime Expect.getSignature("{s}", "<green>expected<r>", false) ++ "\n\n{s}\n";
+            globalObject.throwPretty(Output.prettyFmt(fmt, true), .{ matcher_name, message });
+            return false;
+        } else {
+            const err = globalObject.createErrorInstance("expect.{s}() did not return an object containing a \"message\" method", .{matcher_name});
+            err.put(globalObject, ZigString.static("name"), ZigString.init("InvalidMatcherError").toValueGC(globalObject));
+            globalObject.throwValue(err);
+            return false;
+        }
+    }
+
+    pub fn applyCustomSymmetricMatcher(globalObject: *JSC.JSGlobalObject, callFrame: *JSC.CallFrame) callconv(.C) JSValue {
+        defer globalObject.bunVM().autoGarbageCollect();
+
+        // retrieve the Expect instance (to get the flags)
+        var expect_jsvalue: JSValue = callFrame.this();
+        var expect: *Expect = Expect.fromJS(expect_jsvalue) orelse {
+            globalObject.throw("Called on an invalid this: must be an instance of Expect", .{});
+            return .zero;
+        };
+
+        // retrieve the user-provided matcher function
+        var func: JSValue = callFrame.callee();
+        var matcher_fn = func.get(globalObject, custom_matcher_fn_property) orelse unreachable;
+        if (!matcher_fn.jsType().isFunction()) {
+            globalObject.throw("Internal consistency error: the ExpectCustomMatcher(matcherName) is not a function!", .{});
+            return .zero;
+        }
+        matcher_fn.ensureStillAlive();
+
+        // retrieve the matcher name
+        const matcher_name = matcher_fn.getName(globalObject);
+
+        if (expect.scope.tests.items.len <= expect.test_id) {
+            globalObject.throw("expect.{s}() must be called in a test", .{matcher_name});
+            return .zero;
+        }
+
+        // retrieve the captured expected value
+        var value = Expect.capturedValueGetCached(expect_jsvalue) orelse {
+            globalObject.throw("Internal consistency error: failed to retrieve the captured value", .{});
+            return .zero;
+        };
+        value = Expect.processPromise(expect.flags, globalObject, value, "[[custom matcher]]", "<green>expected<r>", false) orelse return .zero;
+        value.ensureStillAlive();
+
+        active_test_expectation_counter.actual += 1;
+
+        // prepare the args array
+        const args_ptr = callFrame.argumentsPtr();
+        const args_count = callFrame.argumentsCount();
+        var allocator = std.heap.stackFallback(8 * @sizeOf(JSValue), globalObject.allocator());
+        var matcher_args = std.ArrayList(JSValue).initCapacity(allocator.get(), args_count + 1) catch {
```

Comment:
Can you add a test that calls a custom matcher 10,000 times in a loop?

This current implementation will likely cause GC issues & crashes. JSC's GC stack scanning is limited. I kind of think we might need to implement at least part of this in C++ to avoid that situation. 

In C++, I would use a `ArgsList` here

Reply:
Run this test locally several times with 1M calls and didn't crash, I think this is safe as the JSValue arguments are already "pinned" (although I might be wrong):
- For symmetric matchers:
  - First arg (the expected value): stored/owned on the `Expect` instance
  - Rest: pinned on the callFrame
- For asymmetric matchers:
  - First arg (the expected value): comes from a callFrame up above (deepEquals or whatever call)
  - Rest: stored/owned in a JSArray at the `ExpectCustomAsymmetricMatcher` instance
 
Also what difference would using an `ArgList` make? As I got it you still have to allocate an array to put the JSValues, as you need to prepend the expected value for the call